src/utilities/geometry/Polyhedron.hpp

```diff
@@ -43,75 +43,98 @@ namespace model {
-  std::vector<Point3d> vertices;
-  std::string name;
-};
+class Surface3d;
 
 class UTILITIES_API Surface3dEdge
 {
  public:
-  Surface3dEdge(Point3d start, Point3d end, Surface3d firstSurface, size_t firstSurfNum);
+  Surface3dEdge(Point3d start, Point3d end, const Surface3d& firstSurface);
+  Surface3dEdge(Point3d start, Point3d end, std::string t_name, size_t t_surfNum);
 
-  Point3d start() const;
-  Point3d end() const;
+  const Point3d& start() const;
+  const Point3d& end() const;
 
   /// check equality: this uses a tolerance
   bool operator==(const Surface3dEdge& other) const;
 
+  // Asserts that isAlmostEqual3dPt(start, other.end) && isAlmostEqual3dPt(end, other.start)
+  bool reverseEqual(const Surface3dEdge& other) const;
+
   /// check inequality
   bool operator!=(const Surface3dEdge& other) const;
 
   size_t count() const;
 
   size_t firstSurfNum() const;
-  void appendSurface(Surface3d surface);
+  std::string firstSurfaceName() const;
+  const std::vector<size_t>& allSurfNums() const;
+  void appendSurface(const Surface3d& surface);
 
   // Checks whether a Point: is not almost equal to the start and end points, and that isPointOnLineBetweenPoints(start, end, testVertex) is true
   bool containsPoint(const Point3d& testVertex);
 
-  std::vector<Surface3d> allSurfaces() const;
+  void markConflictedOrientation();
+  bool hasConflictedOrientation() const;
+
+  // Indicates this edge has been created and wasn't part of the original surface (cf updateZonePolygonsForMissingColinearPoints)
+  void markCreated();
+  bool hasBeenCreated() const;
+
+  void resetEdgeMatching();
+
+  Vector3d asVector() const;
+
+  // If it containsPoint the testVertex, Split in place the current edge from [start, vertex] vertex and returns the new next [vertex, end]
+  boost::optional<Surface3dEdge> splitEdge(Point3d testVertex);
 
  private:
   Point3d m_start;
   Point3d m_end;
-  size_t m_firstSurfNum;
-  std::vector<Surface3d> m_allSurfaces;
+  bool m_conflictedOrientation = false;
+  bool m_hasBeenCreated = false;
+  std::string m_firstSurfaceName;
+  std::vector<size_t> m_allSurfNums;
 };
 
-using Surface3dEdgeVector = std::vector<Surface3dEdge>;
-
-class VolumeEnclosedReturnType
+// A thin wrapper to improve reporting, by attaching a name to the vertices
+class UTILITIES_API Surface3d
 {
  public:
-  VolumeEnclosedReturnType() = default;
+  Surface3d(std::vector<Point3d> t_vertices, std::string t_name, size_t t_surfNum);
+  std::vector<Point3d> vertices;
+  std::string name;
+  size_t surfNum;
+  std::vector<Surface3dEdge> edges;
+
+  size_t numConflictedEdges() const;
+  double ratioOfConflictedEdges() const;
+
+  // There is no check done whatsoevever to ensure the surface is planar.
+  bool isConvex() const;
 
-  bool isEnclosedVolume = false;
-  Surface3dEdgeVector edgesNot2;
+  void resetEdgeMatching();
+
+  bool operator<(const Surface3d& rhs) const;
 };
 
 class UTILITIES_API Polyhedron
```

Comment:
Where did this come from, is this yours @jmarrec ?

Reply:
The enclosed checks were initially in E+, I've had to modify that portion last year so after learning about polyhedron geometry I decided to tackle the Space::volume issues were still there 13 years later.
The work in this PR is original.